ostd/src/mm/kspace/kva.rs

```diff
@@ -0,0 +1,393 @@
+                to_remove = Some(*key);
+                break;
+            }
+        }
+
+        if let Some(key) = to_remove {
+            if let Some(freenode) = freelist.get_mut(&key) {
+                if freenode.block.end - size == freenode.block.start {
+                    freelist.remove(&key);
+                } else {
+                    freenode.block.end -= size;
+                }
+            }
+        }
+
+        if let Some(range) = allocate_range {
+            Ok(range)
+        } else {
+            Err(Error::KvaAllocError)
+        }
+    }
+
+    /// Frees a kernel virtual area.
+    fn free(&self, range: Range<Vaddr>) {
+        let mut lock_guard = self.freelist.lock();
+        let freelist = lock_guard.as_mut().unwrap_or_else(|| {
+            panic!("Free a 'Kva' when 'VirtAddrAllocator' has not been initialized.")
+        });
+        // 1. get the previous free block, check if we can merge this block with the free one
+        //     - if contiguous, merge this area with the free block.
+        //     - if not contiguous, create a new free block, insert it into the list.
+        let mut free_range = range.clone();
+
+        if let Some((prev_va, prev_node)) = freelist
+            .upper_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_prev()
+        {
+            if prev_node.block.end == free_range.start {
+                let prev_va = *prev_va;
+                free_range.start = prev_node.block.start;
+                freelist.remove(&prev_va);
+            }
+        }
+        freelist.insert(free_range.start, KvaFreeNode::new(free_range.clone()));
+
+        // 2. check if we can merge the current block with the next block, if we can, do so.
+        if let Some((next_va, next_node)) = freelist
+            .lower_bound_mut(core::ops::Bound::Excluded(&free_range.start))
+            .peek_next()
+        {
+            if free_range.end == next_node.block.start {
+                let next_va = *next_va;
+                free_range.end = next_node.block.end;
+                freelist.remove(&next_va);
+                freelist.get_mut(&free_range.start).unwrap().block.end = free_range.end;
+            }
+        }
+    }
+}
+
+pub static KVA_TRACKED_ALLOCATOR: VirtAddrAllocator =
+    VirtAddrAllocator::new(TRACKED_MAPPED_PAGES_RANGE);
+pub static KVA_UNTRACKED_ALLOCATOR: VirtAddrAllocator = VirtAddrAllocator::new(VMALLOC_VADDR_RANGE);
+
+#[derive(Debug)]
+pub struct Tracked;
+#[derive(Debug)]
+pub struct UnTracked;
+
+pub trait AllocatorSelector {
+    fn select_allocator() -> &'static VirtAddrAllocator;
+}
+
+impl AllocatorSelector for Tracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_TRACKED_ALLOCATOR
+    }
+}
+
+impl AllocatorSelector for UnTracked {
+    fn select_allocator() -> &'static VirtAddrAllocator {
+        &KVA_UNTRACKED_ALLOCATOR
+    }
+}
+
+#[derive(Debug)]
+pub struct Kva<M: AllocatorSelector + 'static> {
```

Comment:
Please add some documentations to the structure. It may be useful as an public OSTD API in the future.